File: 903-valid-permutations-for-di-sequence/903-valid-permutations-for-di-sequence.java
class Solution
{
    static int v[];
    static int mod=(int)(1e9+7);
    public int val(String s,int i,int j,int n,int dp[][])
    {
        if(i==n)
        {
            return 1;
        }
        int max=0;
        if(dp[i][j]!=-1)
        {
            return dp[i][j];
        }
        if(s.charAt(i)=='D')
        {
            for(int k=j-1;k>=0;k--)
            {
                if(v[k]==0)
                {
                    v[k]=1;
                    max=(max%mod+val(s,i+1,k,n,dp)%mod)%mod;
                    v[k]=0;
                }
            }
        }
        else
        {
            for(int k=j+1;k<=n;++k)
            {
                if(v[k]==0)
                {
                    v[k]=1;
                    max=(max%mod+val(s,i+1,k,n,dp)%mod)%mod;
                    v[k]=0;
                }
            }
        }
        return dp[i][j]=max;
    }
    public int numPermsDISequence(String s) 
    {
        int n=s.length();
        int dp[][]=new int[n+10][n+10];
        v=new int[n+10];
        for(int i=0;i<=n;i++)
        {
            Arrays.fill(dp[i],-1);   
        }
        int max=0;
        for(int i=0;i<=n;i++)
        {
            v[i]=1;
            max=(max%mod+val(s,0,i,n,dp)%mod)%mod;
            v[i]=0;
            
        }
        return max;
    }
}

File: 903-valid-permutations-for-di-sequence/903-valid-permutations-for-di-sequence.cpp
class Solution {
public:
    
    int mod = 1e9 + 7;
    
    int n;
    
    vector<int> vis;
    
    int dp[202][202];
    
    long long func(int i,int j,string &s){
        if(i == n) return 1;
        if(dp[i][j] != -1) return dp[i][j];
        long long ans=0;
        if(s[i] == 'D'){
            
            for(int k=j-1;k>=0;k--){
                if(vis[k] == false){
                    vis[k] = true;
                    long long tmp = func(i+1,k,s);
                    ans += tmp;
                    ans%=mod;
                    vis[k] = false;
                }
            }
        }else{
            for(int k=j+1;k<=n;k++){
                if(vis[k] == false){
                    vis[k] = true;
                    long long tmp = func(i+1,k,s);
                    ans += tmp;
                    ans%=mod;
                    vis[k] = false;
                }
            }
        }
        
        return dp[i][j] = ans;
        
    }
    
    int numPermsDISequence(string s) {
        
        n = s.size();
        memset(dp,-1,sizeof(dp));
        
        vis = vector<int> (n+1,false);
        
        long long ans = 0;
        for(int i=0;i<=n;i++){
            vis[i] = true;
            long long tmp = func(0,i,s);
            ans += tmp;
            ans%=mod;
            vis[i] = false;
        }
        
        return ans;
        
    }
};